### PR TITLE
[FT] Setup and Main

### PR DESCRIPTION
**Pull request by dhillondeep**
Summary:
1) Added `setup.py` (feel free to modify it because I just learned how to make it)
2) Modified include paths based on `setup.py` installs
3) Added `wcosa.py` as the main script. So far only 3 options are there (boards, create and update)

### CODE REVIEW

File: wcosa/boards.json [hunk 1]
```diff
@@ -1,100 +1,120 @@
 {
     "diecimila": {
+        "name": "Cosa Arduino Diecimila (ATmega168/BOOT_168)",
```
Comment (Mogball) [inline wcosa/boards.json:3]: Just realized now but you can read these directly from `cosa/boards.txt`, is there a reason we are keeping like copies of the config in JSON format?

File: wcosa/handle.py [hunk 2]
```diff
@@ -6,10 +6,10 @@
 
 from shutil import copyfile
 from colorama import Fore
-from wcosa.others.output import write, writeln
-from wcosa.others import helper
-from wcosa.templates import config
-from wcosa.templates import cmake
+from others.output import write, writeln
```
Comment (Mogball) [inline wcosa/handle.py]: I support renaming `others` to `utils`

File: wcosa/handle.py [hunk 3]
```diff
@@ -121,3 +121,7 @@ def update_wcosa(path, board):
         cmake.parse_update(project_path + "/CMakeLists.txt", project_data)
 
     writeln("done")
+
+
+if __name__ == "__main__":
```
Comment (Mogball) [inline wcosa/handle.py]: left in debugging statement?

**Review by Mogball — COMMENTED** [review]
I'll check out the changes in a bit

File: wcosa/boards.json (hunk 1 above)
Comment (dhillondeep) [reply wcosa/boards.json:3]: yeah, we can but that file is a bit hard to parse. So I have a parser which gathers essential information from that file and creates this file. Now we can just use this file. Also I don't have to parse the whole file whenever someone asks for a board

File: wcosa/handle.py (hunk 2 above)
Comment (dhillondeep) [reply wcosa/handle.py]: sure

File: wcosa/handle.py (hunk 3 above)
Comment (dhillondeep) [reply wcosa/handle.py]: yeah oops will fix that

File: wcosa/boards.json (hunk 1 above)
Comment (Mogball) [reply wcosa/boards.json:3]: Fair enough,

**Comment by Mogball** [comment]
1. should add `WCosa.egg-info/` to `.gitignore`
2. I ran `python wcosa.py create --path [project_path] --board [board_name] --ide clion` and the project `wcosa/CMakeLists.txt` was created with `{{ name }}` as empty so I got an error when trying to configure
3. If `--path` is not specified the current calling directory should be used
4. When `--port` is not specified it is filled in as `PORT None` but the option should not be specified at all when there is no port, otherwise an upload target will be created

On that last note `wcosa upload` should autodetect the port because it will change from OS to OS and environment to environment. You can do this using `pyserial`.

Otherwise `create` and `update` seem to work.

@dhillondeep

**Comment by dhillondeep** [comment]
All these updates have been made in my current version, upload and build are ready. Going to push in a moment. Currently, if the path is not specified current directory is being used. The way port works in upload is this:
1) if the port is provided through the command line, use that port,
2) if no port is provided but the port in ` config.py` is not None, use that port
3) If none of the above is true, automatically select one

**Comment by Mogball** [comment]
The port might change during operation (or if the person restarts their PC), so `wcosa upload` should do a check and update the port in `wcosa/CMakeLists.txt`, reconfigure the cmake build, and then run the upload target. The port specified at project configuration/update time shouldn't be set in stone.

**Comment by dhillondeep** [comment]
It performs a lot of checks, yes no port is being specified at the create, update and build time. I am going to push in a bit, let me know how it works on Linux and python 2.7. I have opened another PR with those changes

**Comment by Mogball** [comment]
Should this one be closed?

**Comment by dhillondeep** [comment]
Yeah it can be closed, I will do it
